rfcs/0192-min-capacity-ensures-workers-do-not-get-unnecessarily-killed.md

```diff
@@ -0,0 +1,456 @@
+
+**Current Implementation Issues:**
+- Worker-manager enforces minCapacity by spawning new workers when capacity drops below threshold
+- Generic-worker shuts down after `afterIdleSeconds` regardless of minCapacity requirements
+- Gap exists between worker shutdown and replacement detection/provisioning
+
+## Proposed Solution
+
+### Core Concept: MinCapacity Workers
+
+Workers fulfilling `minCapacity >= 1` requirements should:
+1. **Extended Idle Timeout**: Receive significantly longer idle timeouts to preserve caches
+2. **Priority Replacement**: Be immediately replaced when they fail or become unhealthy
+3. **Per-Variant MinCapacity**: Each pool variant (gecko-1 vs gecko-3) maintains its own minCapacity workers
+
+### Implementation Approach
+
+#### 1. Worker Identification and Tagging
+
+**Database Schema Addition:**
+```sql
+ALTER TABLE workers ADD COLUMN intended_role text; -- 'min-capacity' or 'overflow'
+```
+
+**Current Workers Table Schema:** [workers table definition](https://github.com/taskcluster/taskcluster/blob/d7fbf0ee9e0d93e079cc7ff069eaceecfd7d29ec/db/tables.yml#L154-L168)
+
+**Worker Provisioning Logic:**
+- Worker-manager determines worker role at spawn time based on current pool needs
+- Workers are launched with predetermined `intended_role`
+- Uses existing [`launch_config_id` system](https://github.com/taskcluster/taskcluster/blob/d7fbf0ee9e0d93e079cc7ff069eaceecfd7d29ec/services/worker-manager/src/data.js#L313-L362) to ensure per-variant capacity tracking
+- **MinCapacity Worker Selection**: When capacity planning, prioritize spawning minCapacity workers first
+
+#### 2. Lifecycle-Based Worker Configuration
+
+Workers are assigned minCapacity or overflow roles at launch time and never change configuration during their lifetime. This approach works within TaskCluster's architectural constraints where worker-manager cannot communicate configuration changes to running workers.
+
+**Launch-Time Configuration:**
+```javascript
+// Worker-manager determines worker role at spawn time
+if (workerWillFulfillMinCapacity) {
+  launchConfig.idleTimeoutSecs = pool.config.minCapacityIdleTimeoutSecs;
+  launchConfig.intendedRole = 'min-capacity';
+} else {
+  launchConfig.idleTimeoutSecs = pool.config.worker_config?.shutdown?.afterIdleSeconds || 600;
+  launchConfig.intendedRole = 'overflow';
+}
+```
+
+**Immutable Worker Behavior:**
+- Workers receive their complete configuration at startup
+- No runtime configuration changes
+- Worker role and idle timeout never change during worker lifetime
+- MinCapacity requirements fulfilled through worker replacement, not reconfiguration
+
+#### 3. Pool Configuration
+
+**New Configuration Options:**
+```yaml
+config:
+  minCapacity: 1
+  minCapacityIdleTimeoutSecs: 86400  # 24 hours - enables minCapacity workers
```

Comment:
or it could be `0` as Boris suggested, but the risk will be much higher if someone changes minCapacity to a lower value - we might not be able to kill those "Forever-workers" easily

Reply:
Let's move the conversation to https://github.com/taskcluster/taskcluster-rfcs/pull/192#discussion_r2382428174. Resolving thread.